README.md
# DatacenterDetector
The DatacenterDetector gem is no more. Below is a simple class to perform the same actions, but using a Redis cache via Kredis.

The following Ruby class is a slim wrapper around the [IP Address API](https://github.com/NikolaiT/IP-Address-API).

The number of hits and misses is also tracked.

# A Working Cache

```ruby
require 'kredis'
require 'open-uri'

class IpAddressApi
  # Outside of Rails, configure Kredis like: 
  #   Kredis::Connections.connections[:shared] = Redis.new(url: "redis://localhost:6379/0")
  attr_accessor :prefix, :agent, :cache_only, :ttl

  def initialize(prefix: IpAddressApi.prefix, agent: IpAddressApi.agent, cache_only: false, ttl: 60 * 60 * 24)
    @cache_only = cache_only
    @prefix     = prefix
    @agent      = agent
    @ttl        = ttl 
  end
  
  def key_name(ip) = "#{@prefix}#{ip}"
  def hit_counter  = @hit_counter  ||= Kredis.counter("#{prefix}hits")
  def miss_counter = @miss_counter ||= Kredis.counter("#{prefix}miss")
  
  def stats
    hit  = hit_counter.value
    miss = miss_counter.value

    {hit: hit, miss:, total: hit + miss}
  end

  def hitrate
    hit  = hit_counter.value
    miss = miss_counter.value
    
    hit / (hit + miss).to_f
  end

  def reset_stats
    hit_counter.reset
    miss_counter.reset
  end

  def lookup(ip=nil)
    return {} if ip.nil?

    ip_data = Kredis.json(key_name(ip), expires_in: ttl)
    
    if ip_data.value.blank? && cache_only == true
      miss!
    elsif ip_data.value.blank? && cache_only == false
      miss!
      ip_data.value = JSON.parse( URI.open("https://api.incolumitas.com/?q=#{ip}", "User-Agent" => agent).read)
      ip_data.value ||= {}
    else
      hit!
    end

    return ip_data.value
  rescue OpenURI::HTTPError => e
    puts("IP Address API error looking up IP: #{ip} : #{e.inspect}")
    return {}
  end

  def self.prefix      = 'IpAddressApi_'
  def self.agent       = "Ruby/#{RUBY_VERSION}"
  def self.stats       = IpAddressApi.new.stats
  def self.reset_stats = IpAddressApi.new.reset_stats
  def self.hitrate     = IpAddressApi.new.hit_rate

  def self.lookup(ip=nil, cache_only: false, agent: nil) = IpAddressApi.new(cache_only:, agent:).lookup(ip)

  private
  def hit!         = hit_counter.increment
  def miss!        = miss_counter.increment
end
```

## Usage

You can use an instance or class methods.

```ruby
> result = IpAddressApi.lookup('1.1.1.1')
> result.dig('is_datacenter')
=> false
> result.dig('company', 'name')
=> "APNIC Research and Development"
> result.dig('asn', 'org')
=> "Cloudflare, Inc."

> client = IpAddressApi.new
> client.lookup('52.93.127.126')
> result.is_datacenter
=> true
> result.name
=> "Amazon Technologies Inc."

The cache records its hitrate:

```ruby
> client.hitrate
=> 0.6829268292682927
```
## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
## 20221229
- Deleted this code and replaced it with a tiny class to cache with Kredis in Redis.
## [0.4.4] - 2022-09-05
- When we get an exception accessing the cache, return nil, not []
- Allow setting SQLite's busy_timout by env and default to 500, not 100
## [0.4.2] - 2022-08-22
- "More robust handling for CIDR - sometimes they're returned as x.x.x.x - y.y.y.y"

## [0.4.1] - 2022-08-22
- Honour `force` in DatacenterDetector#query

## [0.4.0] - 2022-08-05
- Store and retreive cache's hitrate.  
- Bug fixes
## [0.3.0] - 2022-08-05
- Store hit / miss data in sqlite to allow a hitrate score
- Handle response with no network found

## [0.2.0] - 2022-08-01

- Add a User Agent
- Update README
- Handle 127.0.0.1 as a special case

## [0.1.0] - 2022-08-01

- Initial release

test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "datacenter_detector"
require "byebug"
require "minitest/autorun"


test/test_datacenter_detector.rb
# frozen_string_literal: true

require "test_helper"

class TestDatacenterDetector < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::DatacenterDetector::VERSION
  end

  def test_loopback_dont_break_it
    ddc_query = DatacenterDetector::Client.new.query(ip: '127.0.0.1')
  end

  def test_it_parses_result
    hash = { cidr: "3.5.140.0/22",
             name: "Amazon AWS",
             is_datacenter: true,
             retreived_at: "CURRENT_TIMESTAMP",
             created_at: "2022-08-01 03:55:00",
             result: { ip: "3.5.140.2",
                       is_datacenter: true,
                       src: "selfPublished",
                       cidr: "3.5.140.0/22",
                       region: "ap-northeast-2",
                       datacenter: "Amazon AWS",
                       service: "AMAZON",
                       network_border_group: "ap-northeast-2",
                       rir: "arin",
                       asn: { asn: 16_509, cidr: "3.5.140.0/22", descr: "AMAZON-02, US", country: "us",
                              active: true },
                       location: { country: "us" },
                       elapsed_ms: 0.21 } }

    to_ostruct = DatacenterDetector.to_ostruct(hash)

    to_hash = DatacenterDetector.to_hash(to_ostruct)

    assert_equal hash, to_hash
  end

  def test_unhashing
    skip
    str = "{\"cidr\":\"3.5.140.0/22\",\"name\":\"Amazon AWS\",\"is_datacenter\":true,\"retreived_at\":\"CURRENT_TIMESTAMP\",\"created_at\":\"2022-08-01 03:55:00\",\"result\":\"#<OpenStruct ip=\\\"3.5.140.2\\\", is_datacenter=true, src=\\\"selfPublished\\\", cidr=\\\"3.5.140.0/22\\\", region=\\\"ap-northeast-2\\\", datacenter=\\\"Amazon AWS\\\", service=\\\"AMAZON\\\", network_border_group=\\\"ap-northeast-2\\\", rir=\\\"arin\\\", asn=\\\"#<OpenStruct asn=16509, cidr=\\\\\\\"3.5.140.0/22\\\\\\\", descr=\\\\\\\"AMAZON-02, US\\\\\\\", country=\\\\\\\"us\\\\\\\", active=true>\\\", location=\\\"#<OpenStruct country=\\\\\\\"us\\\\\\\">\\\", elapsed_ms=0.21>\"}"

    DatacenterDetector.to_hash(JSON.parse(str))
  end

  def test_no_asn_result
    response = {
      status: "200",
      result: {
        ip: "5.255.253.156",
        is_datacenter: false,
        rir: "ripe ncc",
        asn: nil,
        location: {
          country: "ru"
        },
        elapsed_ms: 38.56
      },
      is_datacenter: false,
      name: nil
    }
    response = DatacenterDetector.to_ostruct(response)
    
    cache = DatacenterDetector::Cache.new()

    cache.add(response.result)
  end

  def test_convert_network_range_to_cidr
    test_cases = [
      "195.201.0.0 - 195.201.255.255" => IPAddr.new("195.201.0.0").tap {|i| i.prefix = 16}
    ]
    test_cases.each do |tc|
      range, cidr = tc.first
      assert cidr == DatacenterDetector.to_cidr(range)
    end
    
    [ "119.17.144.0/20", "192.168.0.0/16"
    ].each { |tc| assert IPAddr.new(tc) == DatacenterDetector.to_cidr(tc) }
  end
end


lib/datacenter_detector.rb
# frozen_string_literal: true

require_relative "datacenter_detector/version"
require_relative "datacenter_detector/client"
require_relative "datacenter_detector/cache"
require "open-uri"
require "json"

module DatacenterDetector
  IP_RE = /\d+\.\d+\.\d+\.\d+/
  CIDR_RE = /\d+\.\d+\.\d+\.\d+\/\d+/

  class ServerError < StandardError; end

  def self.query(ip)
    data = URI.open(url(ip), "User-Agent" => "DatacenterDetector/#{::DatacenterDetector::VERSION} (https://github.com/dkam/datacenter_detector)")
    body = data.read
    doc = JSON.parse(body)
    result = { status: data.status[0] }
    response = if doc.is_a?(Hash)
                 DatacenterDetector.to_ostruct(result.merge!({ result: doc }))
               else
                 DatacenterDetector.to_ostruct(result.merge!({ error: body }))
               end
    response.is_datacenter = response.result.is_datacenter
    response.name = response.result.server || response.result.datacenter || response.result.asn&.descr

    if ip == '127.0.0.1'
      response.result.cidr = '127.0.0.1/8'
      response.result.server = 'Loopback'
    end

    response
  rescue OpenURI::HTTPError => e
    raise DatacenterDetector::ServerError, "Message: #{e.message}"
  end

  def self.url(ip)
    "https://api.incolumitas.com/datacenter?ip=#{ip}"
  end

  # Recursively convert an object into an OpenStruct
  def self.to_ostruct(obj)
    obj = JSON.parse(obj) if obj.is_a?(String) && obj.valid_json?

    if obj.is_a?(Hash)
      OpenStruct.new(obj.map { |key, val| [key, to_ostruct(val)] }.to_h)
    elsif obj.is_a?(Array)
      obj.map { |o| to_ostruct(o) }
    else # Assumed to be a primitive value
      obj
    end
  end

  # Recursively convert an OpenStruct into a Hash
  def self.to_hash(obj)
    if obj.is_a?(OpenStruct)
      obj.to_h.transform_values { |v| DatacenterDetector.to_hash(v) }
    else
      obj
    end
  end

  def self.to_cidr(range)
    return IPAddr.new(range) if range.match?(CIDR_RE)

    first, last = range.scan(IP_RE)
    first = IPAddr.new(first)
    last  = IPAddr.new( last)

    first.prefix -= 1 while !first.include?(last)
    first
  end
end

class String
  # https://gist.github.com/ascendbruce/7070951
  def valid_json?
    result = JSON.parse(self)
    result.is_a?(Hash) || result.is_a?(Array)
  rescue JSON::ParserError, TypeError
    false
  end
end


lib/datacenter_detector/cache.rb
require "sqlite3"
module DatacenterDetector
  class Cache
    attr_accessor :hit_count, :miss_count

    def initialize(dbf: Cache.default_database_file)
      @hit_count = 0
      @miss_count = 0
      @db = SQLite3::Database.new dbf || Cache.default_database_file
      @db.busy_timeout = ENV.fetch('sqlite_busy_timout', 500)
      setup_range
      setup_range6
      setup_agents
      setup_cache_counter
    end

    def hitrate
      hit = get_counter('hit')
      miss = get_counter('miss')

      hit / (hit + miss).to_f
    end

    def hit!
      increment('hit')
    end

    def miss!
      increment('miss')
    end

    def increment(counter)
      @db.transaction do |d|
        d.execute(
          "UPDATE cache_counter SET value=value + 1 WHERE name=?", counter
        )
      end
    end

    def get_counter(counter)
      return @db.execute( "select value from cache_counter where name = ?", counter )&.first&.first
    end

    def add(result)
      result = result.result if result.respond_to?(:result)
      cidr = result.cidr || result.asn&.cidr
      return if cidr.nil?
      network = DatacenterDetector.to_cidr(cidr)

      if network.ipv6?
        table  = "ranges6"
        start  = network.to_range.first.to_i.to_s(16)
        finish = network.to_range.last.to_i.to_s(16)
      else
        table  = "ranges"
        start  = network.to_range.first.to_i
        finish = network.to_range.last.to_i
      end

      result_hash = DatacenterDetector.to_hash(result)

      record = [cidr, start, finish, result.is_datacenter ? 1 : 0,
                result.server || result.datacenter || result.asn&.descr, result_hash.to_json]

      @db.transaction do |d|
        d.execute(
          "INSERT INTO #{table} (cidr, start, finish, is_datacenter, name, response, retreived_at, created_at)  VALUES (?, ?, ?, ?, ?, ?, CURRENT_TIMESTAMP, CURRENT_TIMESTAMP)", record
        )
      end
    rescue IPAddr::AddressFamilyError => e
      puts "Exception #{e.inspect}"
      puts "CIDR: #{cidr}"
      puts "Result: #{result}"
    end

    def get(ip:, force: false)
      return nil if force

      a = IPAddr.new(ip)
      resp = if a.ipv6?
               @db.execute(
                 "select cidr, name, is_datacenter, retreived_at, created_at, response from ranges6 where start <= ? and finish >= ?", [
                   a.to_i.to_s(16), a.to_i.to_s(16)
                 ]
               ).first
             else
               @db.execute(
                 "select cidr, name, is_datacenter, retreived_at, created_at, response from ranges where start <= ? and finish >= ?", [
                   a.to_i, a.to_i
                 ]
               ).first
             end

      miss! and return nil if resp.nil?
      hit!

      resp = { cidr: resp[0], name: resp[1], is_datacenter: (resp[2] == 1), retreived_at: resp[3], created_at: resp[4],
               result: resp[5], status: 301 }

      DatacenterDetector.to_ostruct(resp)
    rescue StandardError => e
      puts "Exception finding ip: #{ip}"
      puts e.message
      nil
    end

    def setup_range
      return unless @db.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='ranges'").length == 0

      @db.execute("CREATE TABLE ranges(cidr TEXT, start INTEGER, finish INTEGER, is_datacenter BOOLEAN, name TEXT, country TEXT, retreived_at timestamp, response TEXT, created_at timestamp CURRENT_TIMESTAMP, UNIQUE(cidr) on conflict replace, UNIQUE(start, finish) ON CONFLICT REPLACE);")
      @db.execute("CREATE INDEX 'idx_start'   ON ranges ( 'start' );")
      @db.execute("CREATE INDEX 'idx_finish'  ON ranges ( 'finish' );")
      @db.execute("CREATE INDEX 'idx_isdc' ON ranges ( 'is_datacenter' );")
    end

    def setup_range6
      return unless @db.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='ranges6'").length == 0

      @db.execute("CREATE TABLE ranges6(cidr TEXT, start INTEGER, finish INTEGER, is_datacenter BOOLEAN, name TEXT, country TEXT, retreived_at timestamp, response TEXT, created_at timestamp CURRENT_TIMESTAMP, UNIQUE(cidr) on conflict replace, UNIQUE(start, finish) ON CONFLICT REPLACE);")
      @db.execute("CREATE INDEX 'idx_start6'   ON ranges6 ( 'start' );")
      @db.execute("CREATE INDEX 'idx_finish6'  ON ranges6 ( 'finish' );")
      @db.execute("CREATE INDEX 'idx_isdc6' ON ranges6 ( 'is_datacenter' );")
    end

    def setup_agents
      return unless @db.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='agents'").length == 0

      @db.execute("CREATE TABLE agents(name VARCHAR, full  VARCHAR, quality INTEGER);")
      @db.execute("CREATE UNIQUE INDEX 'uniq_name'  ON agents ( 'name' );")
      @db.execute("CREATE INDEX 'idx_agent_quality' ON agents ( 'quality' );")
    end

    def setup_cache_counter
      return unless @db.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='cache_counter'").length == 0

      @db.execute("CREATE TABLE cache_counter(name VARCHAR, value INTEGER);")
      @db.execute("INSERT INTO cache_counter (name, value)  VALUES ('hit', 0)")
      @db.execute("INSERT INTO cache_counter (name, value)  VALUES ('miss', 0)")
    rescue SQLite3::SQLException => e
      puts "Exception creating cache_counter: #{e.inspect}"
    end

    def self.default_database_file
      if defined?(Rails.root)
        File.join(Rails.root, "db", "datacenter_detector.db")
      else
        File.join(File.expand_path("~"), ".datacenter_detector.db")
      end
    end
  end
end


ip_address_api.rb
require 'kredis'
require 'open-uri'

class IpAddressApi
  # Outside of Rails, configure Kredis like: 
  #   Kredis::Connections.connections[:shared] = Redis.new(url: "redis://localhost:6379/0")
  attr_accessor :prefix, :agent, :cache_only, :ttl

  def initialize(prefix: IpAddressApi.prefix, agent: IpAddressApi.agent, cache_only: false, ttl: 60 * 60 * 24)
    @cache_only = cache_only
    @prefix     = prefix
    @agent      = agent
    @ttl        = ttl 
  end
  
  def key_name(ip) = "#{@prefix}#{ip}"
  def hit_counter  = @hit_counter  ||= Kredis.counter("#{prefix}hits")
  def miss_counter = @miss_counter ||= Kredis.counter("#{prefix}miss")
  
  def stats
    hit  = hit_counter.value
    miss = miss_counter.value

    {hit: hit, miss:, total: hit + miss}
  end

  def hitrate
    hit  = hit_counter.value
    miss = miss_counter.value
    
    hit / (hit + miss).to_f
  end

  def reset_stats
    hit_counter.reset
    miss_counter.reset
  end

  def lookup(ip=nil)
    return {} if ip.nil?

    ip_data = Kredis.json(key_name(ip), expires_in: ttl)
    
    if ip_data.value.blank? && cache_only == true
      miss!
    elsif ip_data.value.blank? && cache_only == false
      miss!
      ip_data.value = JSON.parse( URI.open("https://api.incolumitas.com/?q=#{ip}", "User-Agent" => agent).read)
      ip_data.value ||= {}
    else
      hit!
    end

    return ip_data.value
  rescue OpenURI::HTTPError => e
    puts("IP Address API error looking up IP: #{ip} : #{e.inspect}")
    return {}
  end

  def self.prefix      = 'IpAddressApi_'
  def self.agent       = "Ruby/#{RUBY_VERSION}"
  def self.stats       = IpAddressApi.new.stats
  def self.reset_stats = IpAddressApi.new.reset_stats
  def self.hitrate     = IpAddressApi.new.hit_rate

  def self.lookup(ip=nil, cache_only: false, agent: nil) = IpAddressApi.new(cache_only:, agent:).lookup(ip)

  private
  def hit!         = hit_counter.increment
  def miss!        = miss_counter.increment
end